Data overlay sandbox feature (#3760)
opentripplanner/OpenTripPlanner: docs/sandbox/DataOverlay.md

# Data Overlay

> TODO DataOverlay

Reviewer: 1 or 2 sentence abstract.

## Contact Info

> 

Reviewer: If we need to contact you (future changes), then add enough info here. If you do not want to add your email name and organisation is good enough.

Reviewer: You may also include some text here to give us a hint if you would like to be responsible for all changes to DataOverlay or welcome any community updates. We might change it, even if you are in charge, but we will try to contact you first, then.

## Changelog
- Initial version (December 2021)

## Documentation

> TODO

Reviewer: Copy in some of the text you wrote on the user mailing list and in the issue. Just to give people an idea of what this can be used for and a hint on how to use it.

### Configuration 

```json
// otp-config.json
{ "otpFeatures": { "DataOverlay" : true } }
```

```json
// build-config.json
{ "dataOverlay" : {

>   "fileName": "graphs/data-file.nc4",

Reviewer: We discussed the "fileName" parameter today at the developer meeting. We think this is fine for now, but if someone want to access input data directly from the cloud, then we will have to change this. We will then most likely add a list of "data-overlay" data sources to the OTP storage. This will require the config to be changed, the `fileName` parameter will not be used by OTP any more - it will be ignored. Instead a list of data overlay data source should be added in the `storage` section in the _build-config.json_. We will also make the necessary changes to support multiple input files.

# Data Overlay

> Use grid data in NetCDF format to populate the graph. Also provides custom route endpoint parameters for the data "penalty" and "threshold". This allows route planning to be based on the custom data calculated penalties. Data examples: air qualuty, environmental, and other data types that are tied to certain geographical locations. 

Reviewer 2: ```suggestion
Use grid data in NetCDF format to populate the graph. Also provides custom route endpoint parameters for the data "penalty" and "threshold". This allows route planning to be based on the custom data calculated penalties. Data examples: air quality, environmental, and other data types that are tied to certain geographical locations.
```

## Contact Info
Developed and maintained by <strong>Metatavu OY</strong>, Finland.

Developers:

> <strong>Katja Danilova</strong> - katja.danilova@metatavu.fi\

Reviewer 2: <strong> works but you can surround text with `*` to make text bold.

<strong>Simeon Platonov</strong> - simeon.platonov@metatavu.fi\
<strong>Daniil Smirnov</strong> - daniil.smirnov@metatavu.fi

In case of any questions please contact any of the people above by emails. We would like to continue developing and improving this feature and would love to hear any ideas from the community.

Company email: info@metatavu.fi

## Changelog
- Initial version (December 2021)

## Documentation

> We have been working with OTP since version 1 mainly for producing the Air Quality affected routing for the city of Helsinki, Finland. That project required us to modify the original OTP quite a lot so we didn't propose our efforts for the community.

Reviewer 2: ```suggestion
With the OTP2 release we decided to create a dedicated layer on top of OTP2 which not only leaves the initial structure of the OpenTripPlanner intact, but also brings some additional features for those, who actually need them. This layer's main feature is populating the graph with a grid data (i.e air quality, temperature, humidity, pressure, wind speed and direction, and e.t.c). For this to work two files are required: the actual data file (i.e in NetCDF format) and a .json settings file which describes the contents of the data file. Please refer to the diagram for more information.
```

> 

Reviewer 2: ```suggestion
With the OTP2 release we decided to create a dedicated layer on top of OTP2 which not only leaves the initial structure of the OpenTripPlanner intact, but also brings some additional features for those, who actually need them. This layer's main feature is populating the graph with a grid data (i.e air quality, temperature, humidity, pressure, wind speed and direction, and e.t.c). For this to work two files are required: the actual data file (i.e in NetCDF format) and a .json settings file which describes the contents of the data file. Please refer to the diagram for more information.
```

> With the OTP2 release we decided to create a dedicated layer on top of OTP2 which not only leaves the initial structure of the Open Trip Planner intact, but also brings some additional features for those, who actually need them. This layer's main feature is populating the graph with a grid data (i.e air quality, temperature, humidity, pressure, wind speed and direction, and e.t.c). For this to work two files are required: the actual data file (i.e in NetCDF format) and a .json settings file which describes the contents of the data file. Please refer to the diagram for more information.

Reviewer 2: ```suggestion
With the OTP2 release we decided to create a dedicated layer on top of OTP2 which not only leaves the initial structure of the OpenTripPlanner intact, but also brings some additional features for those, who actually need them. This layer's main feature is populating the graph with a grid data (i.e air quality, temperature, humidity, pressure, wind speed and direction, and e.t.c). For this to work two files are required: the actual data file (i.e in NetCDF format) and a .json settings file which describes the contents of the data file. Please refer to the diagram for more information.
```

It is a sandbox feature.

Please see the configuration part for setup instructions and examples.

### Configuration 

Enable the feature by including it to the ```otp-config.json```:

```json
// otp-config.json
{ "otpFeatures": { "DataOverlay" : true } }
```

> TODO - Katja, please provide other instructions here

Reviewer 2: @kaeline Is this done? Can this TODO be removed?

Reviewer 3: @leonardehrenfried sorry, not yet, I wanted to test it out first.

Reviewer 3: @leonardehrenfried Not done yet, I will add more detailed description of how to set the configuration for the data file
